lib/scidata/caltech101.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Scidata.Caltech101 do
+  @moduledoc """
+  Module for downloading the [Caltech101 dataset](http://www.vision.caltech.edu/Image_Datasets/Caltech101).
+  """
+
+  require Scidata.Utils
+  alias Scidata.Utils
+
+  # NOTE: The original url on the website is "https://drive.google.com/u/0/uc?export=download&confirm=R4MY&id=137RyRjvTBkBiIfeYBNZBtViDHQ6_Ewsp".
+  # However, I was unable to get around the redirection issue when downloading large files from GDrive.
+  @base_url "https://s3.amazonaws.com/fast-ai-imageclas/caltech_101.tgz"
+  @labels_shape {9144, 1}
+
+  @doc """
+  Downloads the Caltech101 training dataset or fetches it locally.
+
+  Returns a tuple of format:
+
+      {{images_binary, images_type, images_shape},
+       {labels_binary, labels_type, labels_shape}}
+
+  If you want to one-hot encode the labels, you can:
+
+      labels_binary
+      |> Nx.from_binary(labels_type)
+      |> Nx.new_axis(-1)
+      |> Nx.equal(Nx.tensor(Enum.to_list(1..102)))
+
+  """
+  def download() do
+    download_dataset(:train)
+  end
+
+  defp download_dataset(_dataset_type) do
+    # Skip first file since it's a temporary file.
+    [_ | files] = Utils.get!(@base_url).body
+
+    {:ok, label_mapping} = get_mapping(:label)
+    {:ok, image_mapping} = get_mapping(:image)
+
+    records =
+      files
+      |> Task.async_stream(&generate_records(&1, label_mapping, image_mapping),
+        max_concurrency: Application.get_env(:scidata, :max_concurrency)
+      )
+      |> Enum.to_list()
+
+    images = Enum.map(records, fn {:ok, record} -> elem(record, 0) end)
+    labels = Enum.map(records, fn {:ok, record} -> elem(record, 1) end)
+    shapes = Enum.map(records, fn {:ok, record} -> elem(record, 2) end)
+
+    {{images, {:u, 8}, shapes}, {labels, {:u, 8}, @labels_shape}}
+  end
+
+  defp get_mapping(:label) do
+    fpath = Path.join([File.cwd!(), "lib/scidata/labels/caltech101_labels.json"])
+    do_get_mapping(fpath)
+  end
+
+  defp get_mapping(:image) do
+    fpath = Path.join([File.cwd!(), "lib/scidata/content/caltech101_image_shapes.json"])
+    do_get_mapping(fpath)
+  end
+
+  defp do_get_mapping(fpath) do
+    with {:ok, body} <- File.read(fpath),
+         {:ok, mapping} <- Jason.decode(body),
+         do: {:ok, mapping}
+  end
+
+  defp generate_records({fname, image}, label_mapping, image_mapping) do
+    [_, class_name, image_fname] =
+      fname
+      |> List.to_string()
+      |> String.downcase()
+      |> String.split("/")
+
+    label = Map.get(label_mapping, class_name)
+    shape = image_mapping |> Map.get("#{label}/#{image_fname}") |> List.to_tuple()
+
+    {image, label, shape}
```

Comment:
For some reason, the byte_size of the binary returned after extracting the tar file using :erl_tar does not match the byte_size returned when we open the file using OpenCv(via evision) and return the binary.

Example:

- For the file `101_ObjectCategories/BACKGROUND_Google/image_0004.jpg`, the image shape is `(200,200,3)`. This record is typically the first returned when we load the dataset using:
```ex
iex> {images, labels} = Scidata.Caltech101.download()
iex> {bin_images, img_type, img_shapes} = images
iex> bin_images |> List.first |> byte_size
10071
iex> img_shape = img_shapes |> List.first
iex> bin_images |> List.first |> Nx.from_binary(img_type) |> Nx.reshape(img_shape)
** (ArgumentError) cannot reshape, current shape {10071} is not compatible with new shape {200, 200, 3}
    (nx 0.1.0-dev) lib/nx/shape.ex:159: Nx.Shape.reshape/2
    (nx 0.1.0-dev) lib/nx.ex:1818: Nx.reshape/3
```

However, reading the file using OpenCV gives the expected shape:

```ex
iex> {:ok, mat} = OpenCV.imread("101_ObjectCategories/BACKGROUND_Google/image_0004.jpg")
iex> {:ok, type} = OpenCV.Mat.type(mat)
iex> {:ok, {rows, cols, channels}} = OpenCV.Mat.shape(mat)
iex> {:ok, binary_data} = OpenCV.Mat.to_binary(mat)
iex> binary_data |> byte_size
120000
iex> binary_data |> Nx.from_binary(img_type) |> Nx.reshape(img_shape)
#Nx.Tensor<
  u8[200][200][3]
  [
    [
      [26, 49, 44],
      [34, 54, 49],
      [29, 41, 35],
      [34, 41, 34],
      [64, 64, 58],
      [64, 65, 55],
      [39, 45, 34],
      [28, 43, 29],
      [58, 83, 69],
      [40, 76, 60],
      [52, 90, 78],
      [51, 91, 80],
      [38, 72, 65],
      [52, 82, 77],
      [63, 83, 84],
      [35, 52, 49],
      [19, 36, ...],
      ...
    ],
    ...
  ]
>
```

Any hints on how to fix this issue? 😅